File: plugin/src/net/sf/orcc/network/transforms/Instantiator.java
/*
 * Copyright (c) 2009, IETR/INSA of Rennes
 * All rights reserved.
 * 
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions are met:
 * 
 *   * Redistributions of source code must retain the above copyright notice,
 *     this list of conditions and the following disclaimer.
 *   * Redistributions in binary form must reproduce the above copyright notice,
 *     this list of conditions and the following disclaimer in the documentation
 *     and/or other materials provided with the distribution.
 *   * Neither the name of the IETR/INSA of Rennes nor the names of its
 *     contributors may be used to endorse or promote products derived from this
 *     software without specific prior written permission.
 * 
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
 * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE
 * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
 * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
 * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
 * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT,
 * STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY
 * WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
 * SUCH DAMAGE.
 */
package net.sf.orcc.network.transforms;

import net.sf.orcc.OrccException;
import net.sf.orcc.common.Port;
import net.sf.orcc.ir.type.IType;
import net.sf.orcc.network.Connection;
import net.sf.orcc.network.Instance;
import net.sf.orcc.network.Network;
import net.sf.orcc.network.Vertex;

import org.jgrapht.DirectedGraph;

/**
 * This class defines a network transformation that instantiates a network.
 * 
 * @author Matthieu Wipliez
 * 
 */
public class Instantiator implements INetworkTransformation {

	private DirectedGraph<Vertex, Connection> graph;

	/**
	 * Walks through the hierarchy, instantiate actors, and checks that
	 * connections actually point to ports defined in actors. Instantiating an
	 * actor implies first loading it and then giving it the right parameters.
	 * 
	 * @throws OrccException
	 *             if an actor could not be instantiated, or a connection is
	 *             wrong
	 */
	public void transform(Network network) throws OrccException {
		graph = network.getGraph();
		for (Vertex vertex : graph.vertexSet()) {
			if (vertex.isInstance()) {
				Instance instance = vertex.getInstance();
				if (instance.isNetwork()) {
					// instantiate the child network
					instance.getNetwork().instantiate();
				} else if (!instance.isBroadcast()) {
					// instantiate the child actor
					instance.instantiate();
				}
			}
		}

		updateConnections();
	}

	/**
	 * Updates the given connection's source and target port by getting the
	 * ports from the source and target instances, after checking the ports
	 * exist and have compatible types.
	 * 
	 * @param connection
	 *            a connection
	 * @throws OrccException
	 */
	private void updateConnection(Connection connection) throws OrccException {
		Vertex srcVertex = graph.getEdgeSource(connection);
		Vertex tgtVertex = graph.getEdgeTarget(connection);

		if (srcVertex.isInstance() && tgtVertex.isInstance()) {
			Instance source = srcVertex.getInstance();
			Instance target = tgtVertex.getInstance();

			String srcPortName = connection.getSource().getName();
			Port srcPort = source.getActor().getOutput(srcPortName);

			String dstPortName = connection.getTarget().getName();
			Port dstPort = target.getActor().getInput(dstPortName);

			// check ports exist
			if (srcPort == null) {
				throw new OrccException("A Connection refers to "
						+ "a non-existent source port: \"" + srcPortName + "\"");
			} else if (dstPort == null) {
				throw new OrccException("A Connection refers to "
						+ "a non-existent target port: \"" + dstPortName + "\"");
			}

			// check port types match
			IType srcType = srcPort.getType();
			IType dstType = dstPort.getType();
			if (!srcType.equals(dstType)) {
				throw new OrccException("Type error: " + source.getActor()
						+ "." + srcPort + " is " + srcType + ", "
						+ target.getActor() + "." + dstPort + " is " + dstType);
			}

			// update connection
			connection.setSource(srcPort);
			connection.setTarget(dstPort);
		}
	}

	/**
	 * Updates the connections of this network. Must be called after actors have
	 * been instantiated.
	 * 
	 * @throws OrccException
	 *             if a connection could not be updated because it references a
	 *             port that does not exist or have source and target ports that
	 *             have incompatible types
	 */
	private void updateConnections() throws OrccException {
		for (Connection connection : graph.edgeSet()) {
			updateConnection(connection);
		}
	}

}

File: projects/Jade/lib/ConfigurationEngine/Instantiator.cpp
/*
 * Copyright (c) 2009, IETR/INSA of Rennes
 * All rights reserved.
 * 
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions are met:
 * 
 *   * Redistributions of source code must retain the above copyright notice,
 *     this list of conditions and the following disclaimer.
 *   * Redistributions in binary form must reproduce the above copyright notice,
 *     this list of conditions and the following disclaimer in the documentation
 *     and/or other materials provided with the distribution.
 *   * Neither the name of the IETR/INSA of Rennes nor the names of its
 *     contributors may be used to endorse or promote products derived from this
 *     software without specific prior written permission.
 * 
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
 * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE
 * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
 * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
 * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
 * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT,
 * STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY
 * WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
 * SUCH DAMAGE.
 */

/**
@brief Implementation of class Instantiator
@author Jerome Gorin
@file Instantiator.cpp
@version 1.0
@date 15/11/2010
*/

//------------------------------
#include <iostream>
#include <string>
#include <stdio.h>

#include "Instantiator.h"

#include "llvm/DerivedTypes.h"

#include "Jade/Configuration/Configuration.h"
#include "Jade/Core/Actor.h"
#include "Jade/Core/Port.h"
#include "Jade/Graph/HDAGGraph.h"
#include "Jade/Core/Network.h"
//------------------------------


using namespace std;
using namespace llvm;

Instantiator::Instantiator(Configuration* configuration){
	this->actors = configuration->getActors();
	this->configuration = configuration;
	Network* network = configuration->getNetwork();
	this->graph = network->getGraph();
	updateInstances();
}


void Instantiator::updateInstances(){
	//Update instances using actors
	map<string, Instance*>::iterator it;
	map<string, Instance*>* instances = configuration->getInstances();
		
	for (it = instances->begin(); it != instances->end(); it++){
		updateInstance(it->second);
	}

	//Update connections using graph
	int edges = graph->getNbEdges();
	for (int i = 0; i < edges; i++){
		updateConnection((Connection*)graph->getEdge(i));
	}
	
}

void Instantiator::updateInstance(Instance* instance){
	map<string, Actor*>::iterator it;

	//Look for the actor using instance clasz
	it = actors->find(instance->getClasz());

	//Actor does not exist
	if (it == actors->end()){
		cerr << "An instance refers to non-existent actor: actor "<< instance->getClasz() << " for instance " << instance->getId();
		exit(0);
	}

	//Set instance to its corresponding actor
	instance->setActor(it->second);
}

void Instantiator::updateConnection(Connection* connection){
	IntegerType* srcPortType;
	string sourceString;
	IntegerType* dstPortType;
	string targetString;
	
	// Get vertex of the connection
	Vertex* srcVertex = (Vertex*)connection->getSource();
	Vertex* dstVertex = (Vertex*)connection->getSink();

	// Update source
	if(srcVertex->isInstance()){
		// Get instance of the connection
		Instance* source = srcVertex->getInstance();
		
		// Get port of the connection 
		Port* srcPortInst = connection->getSourcePort();

		// Get same port from the actor
		Actor* actor = source->getActor();
		Port* srcPort = actor->getOutput(srcPortInst->getName());
		sourceString = srcPort->getName();
		srcPortType = srcPort->getType();

		if (srcPort == NULL){
			cerr << "A Connection refers to non-existent source port: " << srcPortInst->getName() << "of instance " << source->getId();
			exit(0);
		}

		// Set port information
		srcPortInst->setType(srcPortType);
		source->setAsOutput(srcPortInst);
	}

	// Update target 
	if(dstVertex->isInstance()){
		// Get instance of the connection 
		Instance* target = dstVertex->getInstance();
		
		// Get port of the connection 
		Port* dstPortInst = connection->getDestinationPort();

		// Get same port from the actor 
		Actor* actor = target->getActor();
		Port* dstPort = actor->getInput(dstPortInst->getName());
		targetString = dstPort->getName();
		dstPortType = dstPort->getType();

		if (dstPort == NULL){
			cerr << "A Connection refers to non-existent destination port: " << dstPort->getName() << " of instance " << target->getId();
			exit(0);
		}

		// Bound GlobalVariable to port from instance
		dstPortInst->setType(dstPortType);
		target->setAsInput(dstPortInst);
	}

	// check port types match
	IntegerType* srcType = cast<IntegerType>(srcPortType);
	IntegerType* dstType = cast<IntegerType>(dstPortType);
	if (srcType->getBitWidth() != dstType->getBitWidth()) {
		cerr << "Type error: size of port " << sourceString.c_str() << " is " << srcType->getBitWidth() << " and different from port " << targetString.c_str() << " of size " << dstType->getBitWidth() << ".\n";
		exit(0);
	}
	
}